src/com/xilinx/rapidwright/rwroute/GlobalSignalRouting.java

```diff
@@ -0,0 +1,854 @@
+		if(debugPrintClkPIPs) printCLKPIPs(clk);
+		
+		vrouteDown = UltraScaleClockRouting.routeToCentroidHRouteOrVRouteAboveBelowCentroid(clk, centroidHRouteNode, centroid.getNeighborClockRegion(0, 0), false);
+		if(clkDebug) {	
+			System.out.println("GET VROUTE DOWN:     " + vrouteDown);
+		}
+		if(debugPrintClkPIPs) printCLKPIPs(clk);
+		
+		// get two sets of clock regions, pick UP set for sinks in the resions with the same row as the centroid clock region
+		upDownClockRegions(clockRegions, centroid);
+		
+		List<RouteNode> upDownDistLines = new ArrayList<>();
+		if(clkDebug) System.out.println("ROUTE UP VROUTE TO HDISTR LINEs:");
+		List<RouteNode> upLines = routeFromDirectionalVRouteToHorizontalDistributionLines(clk, vrouteUp, upClockRegions, false);
+		if(upLines != null) upDownDistLines.addAll(upLines);
+		
+		if(clkDebug) System.out.println("ROUTE DOWN VROUTE TO HDISTR LINEs:");
+		List<RouteNode> downLines = routeFromDirectionalVRouteToHorizontalDistributionLines(clk, vrouteDown, downClockRegions, true);
+		if(downLines != null) upDownDistLines.addAll(downLines);
+		
+		if(clkDebug) System.out.println("GET LCB PIN MAPPINGS");
+		Map<RouteNode, ArrayList<SitePinInst>> lcbMappings = getLCBPinMappings(clk);
+		
+		if(clkDebug) System.out.println("ROUTE HDISTRs TO LCBs");
+		UltraScaleClockRouting.routeDistributionToLCBs(clk, upDownDistLines, lcbMappings.keySet());		
+		if(debugPrintClkPIPs) printCLKPIPs(clk);
+		
+		if(clkDebug) System.out.println("ROUTE LCB TO SINKS");
+		UltraScaleClockRouting.routeLCBsToSinks(clk, lcbMappings);
+		if(debugPrintClkPIPs) printCLKPIPs(clk);
+		
+		Set<PIP> clkPIPsWithoutDuplication = new HashSet<>();
+		clkPIPsWithoutDuplication.addAll(clk.getPIPs());
+		clk.getPIPs().clear();
+		clk.setPIPs(clkPIPsWithoutDuplication);
+		
+		if(debugPrintClkPIPs) {
+			System.out.println("FINAL CLK ROUTING");
+			printCLKPIPs(clk);
+		}
+	}
+	
+	/**
+	 * Gets clock regions of a net's sink pins.
+	 * @param clk The net in question.
+	 * @return A list of clock regions of the net's sink pins.
+	 */
+	private static List<ClockRegion> getClockRegionsOfNet(Net clk) {
+		List<ClockRegion> clockRegions = new ArrayList<>();
+		for(SitePinInst pin : clk.getPins()) {
+			if(pin.isOutPin()) continue;
+			Tile t = pin.getTile();
+			ClockRegion cr = t.getClockRegion();
+			if(!clockRegions.contains(cr)) clockRegions.add(cr);
+		}
+		return clockRegions;
+	}
+	
+	/**
+	 * Routes from a GLOBAL_VERTICAL_ROUTE to horizontal distribution lines.
+	 * @param clk The clock net to be routed.
+	 * @param vroute The node to start the route.
+	 * @param clockRegions Target clock regions.
+	 * @param down To indicate if is is routing to the group of top clock regions.
+	 * @return A list of RouteNodes indicating the reached horizontal distribution lines.
+	 */
+	private static List<RouteNode> routeFromDirectionalVRouteToHorizontalDistributionLines(Net clk, RouteNode vroute, List<ClockRegion> clockRegions, boolean down) {
```

Comment:
The name of this method seems like it should be in `UltraScaleClockRouting` and should probably be shortened a bit.  How does this method differ from those that route through a centroid as that seems to be what it does?